ion-c-sys-macros/src/lib.rs

```diff
@@ -0,0 +1,84 @@
+use proc_macro::TokenStream;
+use proc_macro2::Span;
+use quote::quote;
+use syn::{parse_quote, FnArg, Ident, PatType, ReturnType};
+
+/// This macro is intended to support the `IonCReaderHandle` type. When any
+/// function returns an `IonCError`, the error should be populated with the
+/// current reader position.
+///
+/// Without this macro, errors could be decorated like so:
+///
+/// ```no_test
+/// fn foo() -> IonCResult<()> {
+///     fn1().map_err(|err| include_current_position(self.reader, err))?
+///     // ...
+///     Ok(())
+/// }
+/// ```
+///
+/// However, functions can fail in many places, so decorating each of these with
+/// `map_err` is tedious and error prone:
+///
+/// ```no_test
+/// fn foo() -> IonCResult<()> {
+///   fn1()?.map_err(..);
+///   fn2()?.map_err(..);
+/// }
+/// ```
+///
+/// This macro rewrites the function such that any error gets the position
+/// added. It does this by moving the existing implementation into a closure and
+/// then calling `map_err` on the result.
+///
+/// To make the variety of possible functions work (which include mutable
+/// borrows, lifetime elision), the closure uses the `move` keyword. There are
+/// some gymnastics (adapted from dtolnay/no-panic) to make that work out.
+#[proc_macro_attribute]
+pub fn position_error(_attr: TokenStream, item: TokenStream) -> TokenStream {
```

Comment:
To unit test this macro in isolation, we have to use proc_macro2, which would require some changes. It's doable, but..

I believe we have excellent coverage. Firstly, the compiler exercises this code when we type `cargo test` in the workspace. It builds ion-c-sys, applies the transforms and if it doesn't compile you'll hear about it. That provides really great coverage because we have to fully implement the reader trait. Next up, we run the normal tests (including things like the loader tests) that exercise the inner implementation still works. Finally, I added a specific integration test in reader.rs that shows position errors show up.